File: 241211/정수 명령 처리 8/process-numeric-commands-8.py
import sys

input = sys.stdin.readline

N = int(input().rstrip())

class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class linkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.node_num = 0

    def push_front(self, data):
        new_node = Node(data)
        new_node.next = self.head

        if self.head != None:
            self.head.prev = new_node
            self.head = new_node
            new_node.prev = None
        else:
            self.head = new_node
            self.tail = new_node
            new_node.prev = None

        self.node_num += 1

    def push_back(self, data):
        new_node = Node(data)
        new_node.prev = self.tail

        if self.tail != None:
            self.tail.next = new_node
            self.tail = new_node
            new_node.next = None
        else:
            self.tail = new_node
            self.head = new_node
            new_node.next = None
        
        self.node_num += 1

    def pop_front(self):
        if self.head == None:
            print("List is empty")

        elif self.head.next == None:
            temp = self.head

            self.head = None
            self.tail = None
            self.node_num = 0
            return temp.data

        else:
            temp = self.head
            temp.next.prev = None
            self.head = temp.next
            temp.next = None
            
            self.node_num -= 1
            return temp.data

    def pop_back(self):
        if self.tail == None:
            print("List is empty")

        elif self.tail.prev == None:
            temp = self.tail
            
            self.head = None
            self.tail = None
            self.node_num = 0
            return temp.data
        
        else:
            temp = self.tail
            temp.prev.next = None
            self.tail = temp.prev
            temp.prev = None

            self.node_num -= 1
            return temp.data
            
    def size(self):
        return self.node_num

    def empty(self):
        if (self.node_num == 0):
            return 1
        else:
            return 0
            
    def front(self):
        if self.head == None:
            print("List is empty")
        else:
            return self.head.data

    def back(self):
        if self.tail == None:
            print("List is empty")
        else:
            return self.tail.data
            
llist = linkedList();
for _ in range(N):
    command = input().rstrip().split(" ")
    
    if command[0] == "push_front":
        llist.push_front(command[1])
    elif command[0] == "push_back":
        llist.push_back(command[1])
    elif command[0] == "pop_front":
        print(llist.pop_front())
    elif command[0] == "pop_back":
        print(llist.pop_back())
    elif command[0] == "size":
        print(llist.size())
    elif command[0] == "empty":
        print(llist.empty())
    elif command[0] == "front":
        print(llist.front())
    elif command[0] == "back":
        print(llist.back())

File: 241211/정수 명령 처리 8/process-numeric-commands-8.java
import java.util.*;
import java.io.*;

public class Main {
    public static class Node{
        Node prev, next;
        int data;
    }
    static Node head, tail;
    static int cnt;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        head = new Node();
        tail = new Node();
        int num = Integer.parseInt(br.readLine());
        for(int i = 0; i < num; i++){
            StringTokenizer st = new StringTokenizer(br.readLine());
            String cmd = st.nextToken();
            if(cmd.equals("push_front")) push_front(Integer.parseInt(st.nextToken()));
            else if(cmd.equals("push_back")) push_back(Integer.parseInt(st.nextToken()));
            else if(cmd.equals("pop_front")) pop_front();
            else if(cmd.equals("pop_back")) pop_back();
            else if(cmd.equals("size")) size();
            else if(cmd.equals("empty")) empty();
            else if(cmd.equals("front")) front();
            else back();
        }
    }

    static void push_front(int data){
        Node node = new Node();
        node.data = data;
        if(cnt == 0){
            head = node;
            tail = node;
        }
        else{
            head.prev = node;
            node.next = head;
            head = node;
        }
        cnt++;
    }

    static void push_back(int data){
        Node node = new Node();
        node.data = data;
        if(cnt == 0){
            head = node;
            tail = node;
        }
        else{
            tail.next = node;
            node.prev = tail;
            tail = node;
        }
        cnt++;
    }

    static void pop_front(){
        int data = head.data;
        if(cnt == 1){
            tail = new Node();
            head = new Node();
        }
        else{
            Node temp = head.next;
            head.next.prev = null;
            head = temp;
        }
        cnt--;
        System.out.println(data);
    }

    static void pop_back(){
        int data =tail.data;
        Node temp = tail.prev;
        if(cnt == 1){
            tail = new Node();
            head = new Node();
        }
        else{
            tail.prev.next = null;
            tail = temp;
        }
        cnt--;
        System.out.println(data);
    }

    static void size(){
        System.out.println(cnt);
    }

    static void empty(){
        System.out.println(cnt == 0 ? 1 : 0);
    }

    static void front(){
        System.out.println(head.data);
    }

    static void back(){
        System.out.println(tail.data);
    }
}